example/undulator/api.py

```diff
@@ -0,0 +1,223 @@
+        self.longitudinal_velocity = self._longitudinal_velocity(self._gg)
+        self.central_wavenumber = 2 * self._gg ** 2
+
+        self.grid = Grid(
+            k_range=tuple(v * self.central_wavenumber for v in [0.02, 1.1]),
+            theta_range=tuple(v / self.mean_lorentz_factor for v in [0.0, 2.0]),
+            phi_range=tuple([0.0, 2.0 * np.pi]),
+            resolution=resolution,
+        )
+        self.spectrum_input = SynchRadInput(
+            grid=self.grid.aslist(),
+            timeStep=self.time_step,
+            dtype=dtype,
+            native=native,
+            ctx=opencl_context,
+        ).asdict()
+
+        self.synch_rad = None
+
+    def _compute_gg(self, lorentz_factor):
+        return lorentz_factor / (1.0 + self.strength ** 2 / 2) ** 0.5
+
+    @staticmethod
+    def _longitudinal_velocity(gg):
+        return (1.0 - gg ** -2) ** 0.5
+
+    def _compute_particle_track(self, lorentz_factor):
+        gg = self._compute_gg(lorentz_factor)
+        longitudinal_velocity = self._longitudinal_velocity(gg)
+
+        z = longitudinal_velocity * self.t
+        ux = self._ux(z - 0.5 * self.time_step)
+        uz = (self.mean_lorentz_factor ** 2 - 1 - ux ** 2) ** 0.5
+        x = (
+            ux[0] / lorentz_factor * self.time_step / 2
+            + np.cumsum(ux / lorentz_factor) * self.time_step
+        )
+        y = self._y(z)
+        uy = self._uy(z)
+
+        return Track(x, y, z, ux, uy, uz)
+
+    def create_particles(self):
+        deviation = self.energy_spread * np.random.randn(self.number_of_particles)
+        lorentz_factors = self.mean_lorentz_factor + deviation
+
+        for lf in lorentz_factors:
+            p = Particle(lorentz_factor=lf, track=self._compute_particle_track(lf))
+            self.particles.append(p)
+
+        return self
+
+    def __getitem__(self, p):
+        return self.particles[p]
+
+    # Calculating the particles orbits (simplified)
+    def _ux(self, z):
+        val = self.strength * np.sin(2 * np.pi * z)
+        val *= (z > 0) * (z < 1.5) * z / 1.5 + (z > 1.5)
+        val *= (z > self.number_of_periods - 1.5) * (z < self.number_of_periods) * (
+            self.number_of_periods - z
+        ) / 1.5 + (z < self.number_of_periods - 1.5)
+        return val
+
+    @staticmethod
+    def _y(z):
+        return np.zeros_like(z)
+
+    @staticmethod
+    def _uy(z):
+        return np.zeros_like(z)
+
+    def energy_analytical(self):
+        return (
+            self.number_of_particles
+            * self.central_wavenumber
+            * J_in_um
+            * (7 * np.pi / 24)
+            / 137.0
+            * self.strength ** 2
+            * (1 + self.strength ** 2 / 2)
+            * self.number_of_periods
+        )
+
+
+def undulator_spectrum(
+    *,
+    number_of_particles=24,
+    resolution=Resolution(k=512, theta=256, phi=36),
+    dtype="float",
+    native=True,
+    opencl_context=(0,),
```

Comment:
I would put `None` for a default context, which provides an interactive search -- for the malti-platform, multi-device environment `(0,)` leads to an error